server.go

```diff
@@ -96,50 +112,94 @@ func (srv *Server) Handler(channel string) http.HandlerFunc {
 			maxConnTimeCh = t.C
 		}
 
+		eventCh := make(chan eventOrComment, srv.BufferSize)
 		sub := &subscription{
 			channel:     channel,
 			lastEventID: req.Header.Get("Last-Event-ID"),
-			out:         make(chan interface{}, srv.BufferSize),
+			out:         eventCh,
 		}
 		srv.subs <- sub
 		flusher := w.(http.Flusher)
 		//nolint: megacheck  // http.CloseNotifier is deprecated, but currently we are retaining compatibility with Go 1.7
 		notifier := w.(http.CloseNotifier)
 		flusher.Flush()
 		enc := NewEncoder(w, useGzip)
+
+		writeEventOrComment := func(ec eventOrComment) bool {
+			if err := enc.Encode(ec); err != nil {
+				srv.unsubs <- sub
+				if srv.Logger != nil {
+					srv.Logger.Println(err)
+				}
+				return false // if this happens, we'll end the handler early because something's clearly broken
+			}
+			flusher.Flush()
+			return true
+		}
+
+		var readMainCh <-chan eventOrComment = eventCh
+		var readBatchCh <-chan Event
+		closedNormally := false
+
+	ReadLoop:
 		for {
 			select {
 			case <-notifier.CloseNotify():
-				srv.unregister <- sub
-				return
+				break ReadLoop
 			case <-maxConnTimeCh: // if MaxConnTime was not set, this is a nil channel and has no effect on the select
-				srv.unregister <- sub // we treat this the same as if the client closed the connection
-				return
-			case ev, ok := <-sub.out:
+				break ReadLoop
+			case ev, ok := <-readMainCh:
```

Comment:
If I understand correctly, this is the most important part of this change. What it is doing (which seems correct to me, but I want to type this out to be sure I understand it fully) is:

* start with the channel that we read main messages to be the one we gave to the producer, and set a nil channel that we could read batch messages from
* check to see if there is a main message, and if so:
  * If the message is actually a batch of messages:
    * temporarily set the channel that we look for main messages on to be `nil` so we don't read anything from it
    * temporarily set the channel that we look for batch messages to the one that we just got in this message on the main channel
  * else, the message is a regular message, so send it
* check to see if there is a message on the batch channel, and if so:
  * send it
  * if the batch channel is empty (but not `nil`), we must have reached the end. Set the batch channel to `nil` and set the main channel that we read from back to the one we gave to the producer so we can start reading live messages again.

Reply:
That's right. Maybe I should spell this out more in code comments.

Reply:
And the motivation for the change is to get away from the way we were doing these "replay" messages before, as described [here](https://github.com/launchdarkly/eventsource/pull/26/files/4e9c862610c12a42ded62297c7f4202c88954ef1#r458467586).